private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,269 @@
+// encoded output of SHAKE256. See golang.org/x/crypto/sha3 and FIPS 202 for
+// details on the SHAKE hash.
+package manifest
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/sha3"
+)
+
+var errPartial error = errors.New("partial record")
+
+const shake256Name = "shake256"
+
+type Digest struct {
+	dtype  string
+	digest []byte
+	hexstr string
+}
+
+func NewDigestFromBytes(dtype string, digest []byte) *Digest {
+	return &Digest{
+		dtype:  dtype,
+		digest: digest,
+		hexstr: hex.EncodeToString(digest),
+	}
+}
+
+func NewDigestFromHex(dtype string, hexstr string) (*Digest, error) {
+	digest, err := hex.DecodeString(hexstr)
+	if err != nil {
+		return nil, err
+	}
+	return NewDigestFromBytes(dtype, digest), nil
+}
+
+func NewDigestFromString(typedDigest string) (*Digest, error) {
+	hashfunc, digestStr, found := strings.Cut(typedDigest, ":")
+	if !found {
+		return nil, errors.New("malformed digest string")
+	}
+	return NewDigestFromHex(hashfunc, digestStr)
+}
+
+func (d *Digest) String() string {
+	return d.Type() + ":" + d.Hex()
+}
+
+func (d *Digest) Type() string {
+	return d.dtype
+}
+
+func (d *Digest) Bytes() []byte {
+	return d.digest
+}
+
+func (d *Digest) Hex() string {
+	return d.hexstr
+}
+
+// ManifestError occurs when a manifest is malformed.
+type ManifestError struct {
+	lineno int
+	msg    string
+}
+
+var _ error = (*ManifestError)(nil)
+
+func newManifestError(lineno int, msg string) *ManifestError {
+	return &ManifestError{
+		lineno: lineno,
+		msg:    msg,
+	}
+}
+
+func (e *ManifestError) Error() string {
+	return fmt.Sprintf("invalid manifest: %d: %s", e.lineno, e.msg)
+}
+
+// Manifest represents a list of paths and their digests.
+type Manifest struct {
+	paths   map[string]*Digest
+	digests map[string]string
+	hash    sha3.ShakeHash
+}
+
+var _ encoding.TextMarshaler = (*Manifest)(nil)
+var _ encoding.TextUnmarshaler = (*Manifest)(nil)
+
+// NewManifest creates an empty manifest.
+func NewManifest() *Manifest {
+	return &Manifest{
+		paths:   make(map[string]*Digest),
+		digests: make(map[string]string),
+		hash:    sha3.NewShake256(),
+	}
+}
+
+func splitManifest(data []byte, atEOF bool) (int, []byte, error) {
+	// Return a line without LF.
+	if i := bytes.IndexByte(data, '\n'); i >= 0 {
+		return i + 1, data[0:i], nil
+	}
+
+	// EOF occurred with a partial line.
+	if atEOF && len(data) != 0 {
+		return 0, nil, errPartial
+	}
+
+	return 0, nil, nil
+}
+
+// NewManifestFromReader builds a manifest an encoded manifest reader.
+func NewManifestFromReader(manifest io.Reader) (*Manifest, error) {
+	m := NewManifest()
+	scanner := bufio.NewScanner(manifest)
+	scanner.Split(splitManifest)
+	lineno := 0
+	for scanner.Scan() {
+		lineno++
+		hash, path, found := strings.Cut(scanner.Text(), "  ")
```

Comment:
What if the path contains multiple spaces?

Reply:
[Related](https://github.com/bufbuild/buf/pull/1499/files#diff-536233f971a641a3a3fc1c6fb13958674f984e202f8dc0cf7e07c0f28ff741eaR18), curious why is it defined with 2 spaces? `<hash type>:<digest>[SP][SP]<path>[LF]`, also curious, why not have `path` first? since it's easier to grok sorted by path instead of by digest

Reply:
> What if the path contains multiple spaces?

`strings.Cut` splits the string at the first occurrence of the separator. `path` can contain as many spaces as needed. It cannot, however, contain a line feed. This is the same restriction Go has for modules and it's a fair restriction in general (so many tools break with files containing newlines). That said, is this something we should support? Do we expect or want to support files with line feeds in them?

> [Related](https://github.com/bufbuild/buf/pull/1499/files#diff-536233f971a641a3a3fc1c6fb13958674f984e202f8dc0cf7e07c0f28ff741eaR18), curious why is it defined with 2 spaces? `<hash type>:<digest>[SP][SP]<path>[LF]`, also curious, why not have `path` first? since it's easier to grok sorted by path when paths are first

*Tradition*. Mostly.

Checksum tooling has long printed the hash before the filename. `cksum` of the ancient POSIX standard uses `<hash>[TAB][<bytes>[TAB]<path>[LF]`. `md5sum` and its family of tools of the GNU world uses `<hash>[SP]<mode><path>[LF]`. And this double space format is what Go uses for modules (and seems inspired by `md5sum` except text vs binary mode is crazy talk, so just space 🌟 ). We can choose any format we want and using the Go format works and looks fine to read.

Technically there's good reasons to put hashes first. First, digests are always the same size. A constant character column is one more way you could parse this manifest. Or to cut by `[SP][SP]` is just going to work (these are illegal characters). `cut -d ' ' -f 3- < manifest` will work and so will `sed 's/^[^ ]* [ \*]//'`, which will work for `md5sum`, Go, and these manifests alike.

Second, there's no path escaping. Placing the path after the hash means you have a fixed set of characters to expect. No need to escape for spaces. The only illegal character is the newline.

Reply:
Thank you for the explanation! It makes total sense, mostly because same-size digests... beauty of monospace fonts! Just one remaining q:

> And this double space format is what Go uses for modules

Is this in `go.mod`, `go.sum` or other file? I see those files use single space `<path>[SP]<version or hash>`

Reply:
It's in the `dirhash` "h1" hashing for go modules. https://pkg.go.dev/golang.org/x/mod@v0.5.1/sumdb/dirhash#Hash1